# Tarea 1 

Alumnos:

* Miguel Espinoza
* Sebastián López
* Pedro Pérez

## P1

### Parte 1

In [1]:
import os

folder = 'data/raw/'  #Carpeta principal en la que estan los archivos
weeks = os.listdir(folder)  #Se agregan las subcarpetas a una lista
weeks

['w13', 'w14', 'w15', 'w16', 'w17']

In [2]:
#Se crean dos listas para guardar los archivos normales
#y los 'furnished'
files = []
files_furnished = []

#Mediante un bucle for se guardan las direcciones de todos
#los archivos en las listas creadas
for week in weeks:
    direccion = folder + week + '/'
    file = os.listdir(direccion)
    files.append(direccion + file[0])
    files_furnished.append(direccion + file[1])
    
files

['data/raw/w13/metrocuadrado_all_w13.csv',
 'data/raw/w14/metrocuadrado_all_w14.csv',
 'data/raw/w15/metrocuadrado_all_w15.csv',
 'data/raw/w16/metrocuadrado_all_w16.csv',
 'data/raw/w17/metrocuadrado_all_w17.csv']

In [3]:
import pandas as pd

#Se lee el primer dataframe para observar los datos
data_all = pd.read_csv(files[0])
data_all.head(3)

property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6  
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2

En primer lugar se combinan los dataframes normales

In [4]:
data_all = pd.read_csv(files[0])

for file in files:
    dato_semanal = pd.read_csv(file)
    data_all = data_all.merge(dato_semanal, how = 'outer') 
    #Se pone 'outer' para hacer la union de todos los dataframes

print('Cantidad de datos: ', len(data_all))
data_all.head(3)

Cantidad de datos:  16470


property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6  
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2

In [5]:
#Se ve si hay duplicados
print('Duplicados: ', data_all.duplicated().sum())

Duplicados:  175


In [6]:
#Se eliminan las filas duplicadas
data_all.drop_duplicates(inplace=True)
print('Cantidad de datos: ', len(data_all))

Cantidad de datos:  16295


Ahora se combinan los archivos furnished

In [7]:
data_furnished = pd.read_csv(files_furnished[0])

for file in files_furnished:
    dato_semanal = pd.read_csv(file)
    data_furnished = data_furnished.merge(dato_semanal, how='outer')

print('Cantidad de datos: ', len(data_furnished))
data_furnished.head(3)

Cantidad de datos:  2101


property_type|rent_type|location       price  n_rooms  \
0  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000      3.0   
1  Casa en Venta Y Arriendo, BELLAVISTA Bogotá D.C..  $1.300.000      3.0   
2  Casa en Arriendo, PORTALES DEL NORTE Bogotá D.C..  $1.800.000      3.0   

   n_bath   surface                                            details  \
0     2.0  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
1     2.0   72.0 m2  hermosa casa en venta y en arriendo en, mandal...   
2     3.0  76.81 m2  Estrato 4. CASA TRIPLEX!! Exterior . Cuenta co...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
1  https://www.metrocuadrado.com/inmueble/venta-c...                 41.4  
2  https://www.metrocuadrado.com/inmueble/arriend...                 41.3

In [8]:
#Se buscan los duplicados y se eliminan
print('Duplicados: ', data_furnished.duplicated().sum())

Duplicados:  2


In [9]:
data_furnished.drop_duplicates(inplace=True)

print('Cantidad de datos: ', len(data_furnished))
#Se observa que son muchos menos los datos de tipo furnished

Cantidad de datos:  2099


Al intentar combinar ambos dataframes usando `data_all.merge(data_furnished, how='outer')` aparece el error:

`ValueError: You are trying to merge on object and float64 columns. If you wish to proceed you should use pd.concat`

Por lo que se decide comprobar los tipos de las columnas

In [10]:
#Se observan las columnas de los dataframe
print('Tipos Data All')
print(data_all.dtypes)
print()
print('Tipos Data Furnished')
print(data_furnished.dtypes)

Tipos Data All
property_type|rent_type|location     object
price                                object
n_rooms                              object
n_bath                               object
surface                              object
details                              object
url                                  object
metrocuadrado_index                 float64
dtype: object

Tipos Data Furnished
property_type|rent_type|location     object
price                                object
n_rooms                             float64
n_bath                              float64
surface                              object
details                              object
url                                  object
metrocuadrado_index                 float64
dtype: object


Se observa que el error se debe a que en `Data All` las columnas `n_rooms` y `n_bath` son de tipo `object`, en cambio en `Data Furnished` son de tipo `float64`.

Este problema se resuelve simplemente cambiando el tipo de esas columnas.

In [11]:
#Se cambian las columnas de 'data_furnished' a object
data_furnished = data_furnished.astype({'n_rooms':'object', 'n_bath':'object'})

#Obs: No se pueden cambiar las columnas de 'data_all' a float64, pues hay un dato llamado '5+'

In [153]:
#Se combinan los dataframes
data = data_all.merge(data_furnished, how='outer', indicator='Furnished')

#Se coloca utiliza el parámetro indicator para indicar de que dataframe 
#proviene cada fila.
#left_only indica que viene del dataframe 'data_all'
#right_only indica que viene del dataframe 'data_furnished'
#y both que proviene de ambos

print('Cantidad de datos: ', len(data))
data.head(3)

Cantidad de datos:  18394


property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

   Furnished  
0  left_only  
1  left_only  
2  left_only

In [154]:
data['Furnished'].replace({'left_only':'No', 'right_only':'Si', 'both':'Ambos'}, inplace=True)

#Se cambian los nombres para indicar más claramente si provienen de furnished o no
data.head(3)

property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished  
0        No  
1        No  
2        No

In [155]:
#Se cuentan la cantidad de cada valores de cada tipo
data['Furnished'].value_counts()

No    16295
Si     2099
Name: Furnished, dtype: int64

Como se observa no hay elementos comunes entre ambos grupos, son disjuntos. No hay archivos con texto 'furnished' que no estén contenidos en archivos con texto 'all'.

In [156]:
#Esto se comprueba tambien haciendo una intersección entre los dataframes
#mediante el argumento 'inner'
pd.merge(data_all, data_furnished, how='inner')

#Se observa que el resultado es un dataframe vacío

Empty DataFrame
Columns: [property_type|rent_type|location, price, n_rooms, n_bath, surface, details, url, metrocuadrado_index]
Index: []

### Parte 2

#### A

Primero se limpia el precio

In [157]:
data['price'].head(3)

0    $1.050.000
1    $3.930.000
2      $750.000
Name: price, dtype: object

Se observa que son de tipo `object`, pues el precio contiene el signo peso '$' y los puntos. Convendría eliminar este signo y los puntos para poder pasar los datos a `float64`.

In [158]:
data['price'] = data['price'].str.replace('$','')
data['price'] = data['price'].str.replace('.','')
data['price'].head(3)

0    1050000
1    3930000
2     750000
Name: price, dtype: object

In [159]:
#Se cambia el tipo a float64
data['price'] = data['price'].astype('float')
data['price'].head(3)

0    1050000.0
1    3930000.0
2     750000.0
Name: price, dtype: float64

Ahora se limpia el área

In [160]:
data['surface'].tail(4)

18390    164.34 m2
18391     110.0 m2
18392     106.0 m2
18393     135.0 m2
Name: surface, dtype: object

Se deben eliminar los carácteres ' m2' al final del número, y luego cambiar el tipo a `float`.

In [161]:
data['surface'] = data['surface'].str.replace(' m2','')
data['surface'] = data['surface'].astype('float')
data['surface'].tail(4)

18390    164.34
18391    110.00
18392    106.00
18393    135.00
Name: surface, dtype: float64

Ahora se limpia el número de habitaciones

In [162]:
data['n_rooms'].head(3)

0    3
1    3
2    3
Name: n_rooms, dtype: object

Parece que fueran números enteros, pero el tipo de la columna es object

In [163]:
#Se ven los valores de la columna
data['n_rooms'].unique()

array(['3', '5', '4', '2', '1', nan, '5+', 3.0, 5.0, 4.0, 2.0, 1.0],
      dtype=object)

Se observa que la columna tiene valores de tipo string y float. Más aún se ve que se podría cambiar inmediatamente el tipo a `float` sino existiera el valor `5+`.

In [164]:
#Se cuenta cuantas entradas hay de cada valor
data['n_rooms'].value_counts()

2      6571
1      6271
3      1351
4      1198
1.0    1040
2.0     938
5       862
3.0      52
5.0      36
4.0      32
5+        9
Name: n_rooms, dtype: int64

Se ve que solo son 9 las entradas con valor `5+`, por lo que es una parte infima del total del dataset que es de $18394$ entradas, simplemente se podrían eliminar. Pero como no se desea botar datos, se cambiará el valor `5+` por $6$, dado que es el entero mayor a 5 más cercano.

In [165]:
data['n_rooms'] = data['n_rooms'].replace('5+',6)
#Se reemplazan el valor que daba problemas, y se muestran los valores de las entradas
data['n_rooms'].unique()

array(['3', '5', '4', '2', '1', nan, 6, 3.0, 5.0, 4.0, 2.0, 1.0],
      dtype=object)

In [166]:
#Como esta correcto, se puede cambiar el tipo de la columna
data['n_rooms'] = data['n_rooms'].astype('float')
data['n_rooms'].head(3)

0    3.0
1    3.0
2    3.0
Name: n_rooms, dtype: float64

Por último se limpia el número de baños

In [167]:
data['n_bath'].tail(3)

18391    3
18392    3
18393    2
Name: n_bath, dtype: object

Nuevamente parecen ser enteros, pero el tipo es object

In [168]:
#Se ven los valores de la columna
data['n_bath'].unique()

array(['2', '1', '4', '3', nan, '5', '5+', 2.0, 3.0, 4.0, 5.0, 1.0],
      dtype=object)

Nuevamente el error viene dado por `5+`, por lo que se procede de igual manera que en `n_rooms`.

In [169]:
data['n_bath'] = data['n_bath'].replace('5+',6) #Se reemplazan '5+' por 6
data['n_bath'] = data['n_bath'].astype('float') #Se cambia el tipo a float
data['n_bath'].head(3)

0    2.0
1    2.0
2    1.0
Name: n_bath, dtype: float64

Finalmente se muestran nuevamente los tipos de las columnas

In [170]:
data.dtypes

property_type|rent_type|location     object
price                               float64
n_rooms                             float64
n_bath                              float64
surface                             float64
details                              object
url                                  object
metrocuadrado_index                 float64
Furnished                            object
dtype: object

#### B

Se debe dividir la columna `property_type|rent_type|location` en 3 columnas distintas.

In [171]:
#Se hace una copia de la data por si se realiza un error
copia_seguridad = data.copy()

In [151]:
#Cuando algo salga mal correr esta celda
data = copia_seguridad.copy()

In [172]:
#Se muestran unos pocos datos de esta columna
data['property_type|rent_type|location']

0          Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..
1        Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...
2        Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...
3              Casa en Arriendo, PARIS GAITAN Bogotá D.C..
4        Casa en Arriendo, BALCONES DE ORIENTE Bogotá D...
                               ...                        
18389    Apartamento en Arriendo, CHICO NORTE Bogotá D.C..
18390    Apartamento en Arriendo, CHICO NORTE II Bogotá...
18391    Apartamento en Venta Y Arriendo, CHICO NORTE B...
18392    Apartamento en Arriendo, MOLINOS NORTE Bogotá ...
18393       Apartamento en Arriendo, EL NOGAL Bogotá D.C..
Name: property_type|rent_type|location, Length: 18394, dtype: object

Se pueden deducir varias cosas:
1. El tipo de propiedad es la primera palabra del string.
2. El tipo de renta aparece después de la palabra 'en' y un espacio ' '.
3. El lugar esta después de la coma.

En primer lugar se decide dividir el string en dos partes separadas por la coma.

In [173]:
new_columns = data['property_type|rent_type|location'].str.split(pat=',', expand=True)
new_columns.head(3)

0                                           1
0  Casa en Arriendo               LA SOLEDAD NORTE Bogotá D.C..
1  Casa en Arriendo    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..
2  Casa en Arriendo   VILLA DE LOS ALPES II SECTOR Bogotá D.C..

In [174]:
#Se insertan las columnas en la data 
data[['property_type|rent_type', 'location']] = new_columns

#Se elimina la columna original
data.drop('property_type|rent_type|location', axis=1, inplace=True)

data.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished property_type|rent_type  \
0        No        Casa en Arriendo   
1        No        Casa en Arriendo   
2        No        Casa en Arriendo   

                                     location  
0               LA SOLEDAD NORTE Bogotá D.C..  
1    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..  
2   VILLA DE LOS ALPES II SECTOR Bogotá D.C..

Ahora se divide la columna `property_type|rent_type` en dos columnas más.

In [175]:
new_columns = data['property_type|rent_type'].str.split(pat='en ', expand=True)
new_columns.head(3)

0         1
0  Casa   Arriendo
1  Casa   Arriendo
2  Casa   Arriendo

In [176]:
#Se insertan las columnas en la data y se elimina la columna original
data[['property','rent']] = new_columns
data.drop('property_type|rent_type', axis=1, inplace=True)
data.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished                                    location property      rent  
0        No               LA SOLEDAD NORTE Bogotá D.C..    Casa   Arriendo  
1        No    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..    Casa   Arriendo  
2        No   VILLA DE LOS ALPES II SECTOR Bogotá D.C..    Casa   Arriendo

Por último se revisa que no hayan espacios vacíos al inicio o final de las nuevas columnas.

In [178]:
print(data['location'][0])
print(data['property'][0])
print(data['rent'][0])

 LA SOLEDAD NORTE Bogotá D.C..
Casa 
Arriendo


Tanto location como property tienen espacios, por lo que se utiliza strip para eliminarlos.

In [182]:
data['location'] = data['location'].str.strip()
data['property'] = data['property'].str.strip()

### Parte 3

### Parte 4

### Parte 5

### Parte 6

## P2

# P3 

## Implementación

### Parte 1

In [86]:
from sklearn.base import BaseEstimator as BE
from sklearn.base import RegressorMixin as RM
import numpy as np

class RegresionBayesianaEmpirica(BE,RM):
    
    def __init__(self,alpha,beta,tol=1e-11,maxiter=200):
        self.set_params(alpha=alpha,beta=beta,tol=tol,maxiter=maxiter)
        
    def get_posteriori(self,X,y,alpha,beta):
        N,d = X.shape
        try:
            S_N = np.linalg.inv(alpha*np.eye(d) + beta*np.dot(X.T,X)) # dim dxd
        except LinAlgError:
            print('Matriz S_N no invertible')
        m_N = beta*np.dot(S_N,np.dot(X.T,y)) # dim dx1

        return m_N, S_N
    
    
    def fit(self,X,y,i=0,verbose=True):
        i += 1
        if verbose:
            print('Iteracion',i)
            print('Valores actuales: alpha=',self.alpha,' beta=',self.beta)
        if i <= self.maxiter:
            N,d = X.shape
            eigen_val, eigen_vec = np.linalg.eig(self.beta*np.dot(X.T,X))
            eigen_val = np.real_if_close(eigen_val,tol=1)
            gamma = np.sum(eigen_val/(self.alpha + eigen_val))
            m_N, S_N = self.get_posteriori(X,y,self.alpha,self.beta)
        
            new_alpha = gamma/(np.dot(m_N.T,m_N))
            new_beta = (N-gamma)/(np.sum((y-np.dot(X,m_N))**2))
        
            if np.abs(self.alpha - new_alpha) < self.tol and np.abs(self.beta - new_beta) < self.tol:
                self.alpha = new_alpha
                self.beta = new_beta
                self.m_N = m_N
                self.S_N = S_N
                print('Valores finales: alpha=',self.alpha,' beta=',self.beta)
            else:
                self.alpha = new_alpha
                self.beta = new_beta
                self.m_N = m_N
                self.S_N = S_N
                self.fit(X,y,i)
                
    
    def predict(self,X_,return_std=False):
        y_ = np.dot(X_,self.m_N)
        y_std = np.sqrt((1/self.beta) + np.diag(np.dot(X_,np.dot(self.S_N,X_.T))))
        
        if return_std:
            return (y_,y_std)
        return y_

### Parte 2

In [87]:
import pandas as pd
import numpy as np

#Datos para prueba
df = pd.read_csv('data.csv')

In [88]:
df.replace([np.inf, -np.inf], np.nan, inplace=True)
df.dropna(inplace=True)
df.drop(['price','surface','url','details'],axis=1,inplace=True)

In [89]:
df_y = df['Precio m2']#.to_numpy()
df_X = df.drop('Precio m2',axis=1)#.to_numpy()

num_cols = df_X.select_dtypes(include=['int64', 'float64']).columns
cat_cols = df_X.select_dtypes(include=['object']).columns

In [90]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder, PolynomialFeatures, FunctionTransformer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

#cat pipe
cat_pipe = Pipeline(steps=[('ordinal', OneHotEncoder(handle_unknown='ignore'))])

#num pipe
#num_pipe = Pipeline(steps=[('poly',PolynomialFeatures(degree=3)),('scaler',StandardScaler())])
num_pipe = Pipeline(steps=[('scaler',StandardScaler()),('poly',PolynomialFeatures(degree=3))])

#Se diferencian columnas numéricas y categóricas
prep = ColumnTransformer(transformers=[('num',num_pipe,num_cols),('cat',cat_pipe,cat_cols)])

### Parte 3

In [93]:
random_state = 7
def run(df_X,df_y):
    from sklearn.model_selection import train_test_split
    import numpy as np
    import pickle
    
    alpha_grid = np.linspace(1e-10,1e-5,2,endpoint=True)
    beta_grid = np.linspace(1e-10,1e-5,2,endpoint=True)
    
    actual_R2 = 0
    actual_RMSE = 0
    
    X_train, X_test, y_train, y_test = train_test_split(df_X, df_y, test_size=0.2, random_state=random_state)
    def run_single(pipeline):
        pipeline.fit(X_train,y_train)
        print('Fit terminado', '\n')
        y_pred = pipeline.predict(X_test) 

        R_2 = pipeline.score(X_test, y_test)
        RMSE = np.sqrt(np.mean((y_pred-y_test)**2))
        return R_2, RMSE
    
    best_alpha = alpha_grid[0]
    best_beta = beta_grid[0]
    for alpha in alpha_grid:
        for beta in beta_grid:
            Tubo = Pipeline(steps=[('prep',prep),('clf', RegresionBayesianaEmpirica(alpha,beta))])
            print("Empezando fit: alpha y beta iniciales",[alpha,beta])
            R_2, RMSE = run_single(Tubo)
            if R_2 > actual_R2:
                actual_R2 = R_2
                actual_RMSE = RMSE
                best_alpha = alpha
                best_beta = beta
                
    best_fit = Pipeline(steps=[('prep',prep),('clf', RegresionBayesianaEmpirica(best_alpha,best_beta))])
    best_fit.fit(X_train,y_train,clf__verbose=False)
    filename = 'modelo.pkl'
    pickle.dump(best_fit,open(filename,'wb'))
    
    print('RMSE:',actual_RMSE,' R_2:',actual_R2)

In [110]:
X_train, X_test, y_train, y_test = train_test_split(df_X, df_y, test_size=0.2, random_state=random_state)
pipa = Pipeline(steps=[('prep',prep),('clf', RegresionBayesianaEmpirica(3.081438002327972e+105,3.9584154802360893e-13,tol=1e10000))])
pipa.fit(X_train,y_train)
pipa.score(X_test,y_test)

Iteracion 1
Valores actuales: alpha= 3.081438002327972e+105  beta= 3.9584154802360893e-13
Valores finales: alpha= 1.205151572169101e+106  beta= 3.9584154802360893e-13


-0.004124334079621317

In [94]:
run(df_X,df_y)

Empezando fit: alpha y beta iniciales [1e-10, 1e-10]
Iteracion 1
Valores actuales: alpha= 1e-10  beta= 1e-10
Iteracion 2
Valores actuales: alpha= 2.181697659789909e-11  beta= 3.882606325465473e-13
Iteracion 3
Valores actuales: alpha= 2.7492693360468703e-10  beta= 3.942406157933409e-13
Iteracion 4
Valores actuales: alpha= 1.5001457759499476e-09  beta= 3.9560178063966053e-13
Iteracion 5
Valores actuales: alpha= 6.687309200686858e-09  beta= 3.9549270448128894e-13
Iteracion 6
Valores actuales: alpha= 2.533826622554281e-08  beta= 3.9569015105353945e-13
Iteracion 7
Valores actuales: alpha= 8.268613906194227e-08  beta= 3.9588498946100757e-13
Iteracion 8
Valores actuales: alpha= 2.2936789906273126e-07  beta= 3.9598961709718165e-13
Iteracion 9
Valores actuales: alpha= 5.452427360918855e-07  beta= 3.9599432508132046e-13
Iteracion 10
Valores actuales: alpha= 1.181849644902296e-06  beta= 3.959525204565529e-13
Iteracion 11
Valores actuales: alpha= 2.4421148458081802e-06  beta= 3.959118980965121e-13

Iteracion 94
Valores actuales: alpha= 1.2999062568356976e+42  beta= 3.9584154802360893e-13
Iteracion 95
Valores actuales: alpha= 5.083938303851856e+42  beta= 3.9584154802360893e-13
Iteracion 96
Valores actuales: alpha= 1.9883302000783402e+43  beta= 3.9584154802360893e-13
Iteracion 97
Valores actuales: alpha= 7.7763669585609e+43  beta= 3.9584154802360893e-13
Iteracion 98
Valores actuales: alpha= 3.0413400687579523e+44  beta= 3.9584154802360893e-13
Iteracion 99
Valores actuales: alpha= 1.189469255132012e+45  beta= 3.9584154802360893e-13
Iteracion 100
Valores actuales: alpha= 4.6520187710613583e+45  beta= 3.9584154802360893e-13
Iteracion 101
Valores actuales: alpha= 1.8194063068830974e+46  beta= 3.9584154802360893e-13
Iteracion 102
Valores actuales: alpha= 7.115704971179126e+46  beta= 3.9584154802360893e-13
Iteracion 103
Valores actuales: alpha= 2.7829549147603713e+47  beta= 3.9584154802360893e-13
Iteracion 104
Valores actuales: alpha= 1.0884147233419553e+48  beta= 3.9584154802360893e-13


Iteracion 184
Valores actuales: alpha= 2.6293802635217053e+95  beta= 3.9584154802360893e-13
Iteracion 185
Valores actuales: alpha= 1.0283516189582963e+96  beta= 3.9584154802360893e-13
Iteracion 186
Valores actuales: alpha= 4.021887084516862e+96  beta= 3.9584154802360893e-13
Iteracion 187
Valores actuales: alpha= 1.5729615651297514e+97  beta= 3.9584154802360893e-13
Iteracion 188
Valores actuales: alpha= 6.151858650881678e+97  beta= 3.9584154802360893e-13
Iteracion 189
Valores actuales: alpha= 2.405994253095812e+98  beta= 3.9584154802360893e-13
Iteracion 190
Valores actuales: alpha= 9.40985265501903e+98  beta= 3.9584154802360893e-13
Iteracion 191
Valores actuales: alpha= 3.680196944578595e+99  beta= 3.9584154802360893e-13
Iteracion 192
Valores actuales: alpha= 1.439326421722619e+100  beta= 3.9584154802360893e-13
Iteracion 193
Valores actuales: alpha= 5.629211097848069e+100  beta= 3.9584154802360893e-13
Iteracion 194
Valores actuales: alpha= 2.201586596750647e+101  beta= 3.958415480236089

Iteracion 75
Valores actuales: alpha= 3.51690567478242e+30  beta= 3.9584154802360893e-13
Iteracion 76
Valores actuales: alpha= 1.3754631441334948e+31  beta= 3.9584154802360893e-13
Iteracion 77
Valores actuales: alpha= 5.379441576824903e+31  beta= 3.9584154802360893e-13
Iteracion 78
Valores actuales: alpha= 2.1039016422866652e+32  beta= 3.9584154802360893e-13
Iteracion 79
Valores actuales: alpha= 8.228367307651132e+32  beta= 3.9584154802360893e-13
Iteracion 80
Valores actuales: alpha= 3.2181175768290355e+33  beta= 3.9584154802360893e-13
Iteracion 81
Valores actuales: alpha= 1.2586070056287124e+34  beta= 3.9584154802360893e-13
Iteracion 82
Valores actuales: alpha= 4.922416775643589e+34  beta= 3.9584154802360893e-13
Iteracion 83
Valores actuales: alpha= 1.925159069095894e+35  beta= 3.9584154802360893e-13
Iteracion 84
Valores actuales: alpha= 7.529304425543264e+35  beta= 3.9584154802360893e-13
Iteracion 85
Valores actuales: alpha= 2.94471381833028e+36  beta= 3.9584154802360893e-13
Iteracio

Iteracion 165
Valores actuales: alpha= 7.113807108253039e+83  beta= 3.9584154802360893e-13
Iteracion 166
Valores actuales: alpha= 2.7822126598497084e+84  beta= 3.9584154802360893e-13
Iteracion 167
Valores actuales: alpha= 1.0881244271646968e+85  beta= 3.9584154802360893e-13
Iteracion 168
Valores actuales: alpha= 4.255658764260165e+85  beta= 3.9584154802360893e-13
Iteracion 169
Valores actuales: alpha= 1.6643897577977228e+86  beta= 3.9584154802360893e-13
Iteracion 170
Valores actuales: alpha= 6.509434659391806e+86  beta= 3.9584154802360893e-13
Iteracion 171
Valores actuales: alpha= 2.545842365730358e+87  beta= 3.9584154802360893e-13
Iteracion 172
Valores actuales: alpha= 9.956799154280335e+87  beta= 3.9584154802360893e-13
Iteracion 173
Valores actuales: alpha= 3.8941079280152794e+88  beta= 3.9584154802360893e-13
Iteracion 174
Valores actuales: alpha= 1.5229870885275976e+89  beta= 3.9584154802360893e-13
Iteracion 175
Valores actuales: alpha= 5.956408283228937e+89  beta= 3.958415480236089

Iteracion 56
Valores actuales: alpha= 6.074417337007939e+22  beta= 3.9584154802360893e-13
Iteracion 57
Valores actuales: alpha= 2.3757069258494845e+23  beta= 3.9584154802360893e-13
Iteracion 58
Valores actuales: alpha= 9.291398803213033e+23  beta= 3.9584154802360893e-13
Iteracion 59
Valores actuales: alpha= 3.633869598181999e+24  beta= 3.9584154802360893e-13
Iteracion 60
Valores actuales: alpha= 1.4212077789648871e+25  beta= 3.9584154802360893e-13
Iteracion 61
Valores actuales: alpha= 5.558349017259222e+25  beta= 3.9584154802360893e-13
Iteracion 62
Valores actuales: alpha= 2.1738724101389725e+26  beta= 3.9584154802360893e-13
Iteracion 63
Valores actuales: alpha= 8.502023246272579e+26  beta= 3.9584154802360893e-13
Iteracion 64
Valores actuales: alpha= 3.325144518281011e+27  beta= 3.9584154802360893e-13
Iteracion 65
Valores actuales: alpha= 1.3004652830491419e+28  beta= 3.9584154802360893e-13
Iteracion 66
Valores actuales: alpha= 5.086124657494237e+28  beta= 3.9584154802360893e-13
Iterac

Iteracion 147
Valores actuales: alpha= 4.805451831539259e+76  beta= 3.9584154802360893e-13
Iteracion 148
Valores actuales: alpha= 1.8794140350664858e+77  beta= 3.9584154802360893e-13
Iteracion 149
Valores actuales: alpha= 7.350395423843995e+77  beta= 3.9584154802360893e-13
Iteracion 150
Valores actuales: alpha= 2.8747424398666583e+78  beta= 3.9584154802360893e-13
Iteracion 151
Valores actuales: alpha= 1.1243128592459666e+79  beta= 3.9584154802360893e-13
Iteracion 152
Valores actuales: alpha= 4.397191859471324e+79  beta= 3.9584154802360893e-13
Iteracion 153
Valores actuales: alpha= 1.7197434050490465e+80  beta= 3.9584154802360893e-13
Iteracion 154
Valores actuales: alpha= 6.725922983868329e+80  beta= 3.9584154802360893e-13
Iteracion 155
Valores actuales: alpha= 2.6305110315941625e+81  beta= 3.9584154802360893e-13
Iteracion 156
Valores actuales: alpha= 1.0287938627805803e+82  beta= 3.9584154802360893e-13
Iteracion 157
Valores actuales: alpha= 4.023616701784206e+82  beta= 3.95841548023608

Iteracion 37
Valores actuales: alpha= 225674507.44510406  beta= 3.9584154802360893e-13
Iteracion 38
Valores actuales: alpha= 882613855.0911918  beta= 3.9584154802360893e-13
Iteracion 39
Valores actuales: alpha= 3451906136.932664  beta= 3.9584154802360893e-13
Iteracion 40
Valores actuales: alpha= 13500418002.12003  beta= 3.9584154802360893e-13
Iteracion 41
Valores actuales: alpha= 52800186042.6957  beta= 3.9584154802360893e-13
Iteracion 42
Valores actuales: alpha= 206501727998.7548  beta= 3.9584154802360893e-13
Iteracion 43
Valores actuales: alpha= 807629041912.6445  beta= 3.9584154802360893e-13
Iteracion 44
Valores actuales: alpha= 3158640248011.24  beta= 3.9584154802360893e-13
Iteracion 45
Valores actuales: alpha= 12353453997554.05  beta= 3.9584154802360893e-13
Iteracion 46
Valores actuales: alpha= 48314405467912.945  beta= 3.9584154802360893e-13
Iteracion 47
Valores actuales: alpha= 188957823146472.03  beta= 3.9584154802360893e-13
Iteracion 48
Valores actuales: alpha= 739014763453235

Iteracion 128
Valores actuales: alpha= 1.7853037006966434e+62  beta= 3.9584154802360893e-13
Iteracion 129
Valores actuales: alpha= 6.982329548958664e+62  beta= 3.9584154802360893e-13
Iteracion 130
Valores actuales: alpha= 2.7307917365116876e+63  beta= 3.9584154802360893e-13
Iteracion 131
Valores actuales: alpha= 1.0680136845320738e+64  beta= 3.9584154802360893e-13
Iteracion 132
Valores actuales: alpha= 4.1770055735002553e+64  beta= 3.9584154802360893e-13
Iteracion 133
Valores actuales: alpha= 1.633628465041286e+65  beta= 3.9584154802360893e-13
Iteracion 134
Valores actuales: alpha= 6.389127126676994e+65  beta= 3.9584154802360893e-13
Iteracion 135
Valores actuales: alpha= 2.498790044026819e+66  beta= 3.9584154802360893e-13
Iteracion 136
Valores actuales: alpha= 9.772777345526158e+66  beta= 3.9584154802360893e-13
Iteracion 137
Valores actuales: alpha= 3.822136928772086e+67  beta= 3.9584154802360893e-13
Iteracion 138
Valores actuales: alpha= 1.4948392034093551e+68  beta= 3.958415480236089

Iteracion 18
Valores actuales: alpha= 0.0012974553784983085  beta= 3.958388058529087e-13
Iteracion 19
Valores actuales: alpha= 0.004962984782707513  beta= 3.9584067540769366e-13
Iteracion 20
Valores actuales: alpha= 0.019296180852395487  beta= 3.958413101146368e-13
Iteracion 21
Valores actuales: alpha= 0.07535263711276799  beta= 3.9584148614475384e-13
Iteracion 22
Valores actuales: alpha= 0.2945894616100979  beta= 3.958415321318666e-13
Iteracion 23
Valores actuales: alpha= 1.1520254658478264  beta= 3.958415439556655e-13
Iteracion 24
Valores actuales: alpha= 4.505460920148745  beta= 3.958415469831799e-13
Iteracion 25
Valores actuales: alpha= 17.620759705614358  beta= 3.958415477575632e-13
Iteracion 26
Valores actuales: alpha= 68.91474302957094  beta= 3.958415479555827e-13
Iteracion 27
Valores actuales: alpha= 269.5257064788317  beta= 3.958415480062153e-13
Iteracion 28
Valores actuales: alpha= 1054.1159458818893  beta= 3.958415480191616e-13
Iteracion 29
Valores actuales: alpha= 4122.6513

Iteracion 110
Valores actuales: alpha= 3.895146820826721e+51  beta= 3.9584154802360893e-13
Iteracion 111
Valores actuales: alpha= 1.52339339990043e+52  beta= 3.9584154802360893e-13
Iteracion 112
Valores actuales: alpha= 5.957997368550107e+52  beta= 3.9584154802360893e-13
Iteracion 113
Valores actuales: alpha= 2.3301750320022492e+53  beta= 3.9584154802360893e-13
Iteracion 114
Valores actuales: alpha= 9.113323393575139e+53  beta= 3.9584154802360893e-13
Iteracion 115
Valores actuales: alpha= 3.564224237889944e+54  beta= 3.9584154802360893e-13
Iteracion 116
Valores actuales: alpha= 1.393969452123054e+55  beta= 3.9584154802360893e-13
Iteracion 117
Valores actuales: alpha= 5.451819817606686e+55  beta= 3.9584154802360893e-13
Iteracion 118
Valores actuales: alpha= 2.132208799725206e+56  beta= 3.9584154802360893e-13
Iteracion 119
Valores actuales: alpha= 8.339076707823784e+56  beta= 3.9584154802360893e-13
Iteracion 120
Valores actuales: alpha= 3.2614160652526774e+57  beta= 3.9584154802360893e-1

Iteracion 200
Valores actuales: alpha= 7.878892897348078e+104  beta= 3.9584154802360893e-13
Iteracion 201
Valores actuales: alpha= 3.081438002327972e+105  beta= 3.9584154802360893e-13
RMSE: 0  R_2: 0


### Parte 4

In [ ]:
#Usar los datos con las columnas EDA, repetir parte 3

### Parte 5

In [95]:
from sklearn.linear_model import BayesianRidge
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df_X, df_y, test_size=0.2, random_state=random_state)
BR_pipe = Pipeline([('prep',prep),('clf',BayesianRidge())])
BR_pipe.fit(X_train,y_train)
print('Fit terminado', '\n')
y_pred = BR_pipe.predict(X_test) 

R_2 = BR_pipe.score(X_test, y_test)
RMSE = np.sqrt(np.mean((y_pred-y_test)**2))
print('RMSE = ',RMSE,' R2 =',R_2,)

Fit terminado 

RMSE =  1681279.9902968705  R2 = -1.45644959157476e-05
